### PR TITLE
BLOCKED (Needs FB): Enable iterative/delta manifests

### PR DESCRIPTION
**Pull request by jwakre**
The master_new_manifests branch has been rebased onto master. This PR applies the updated patches to the master branch which enables iterative/delta manifests.

### CODE REVIEW

**Comment by jwakre** [comment]
The last 3 commits were not part of the original iterative/delta manifest PR. I added them to resolve issues in the rebase while also making it easier to review.

File: src/manifest.c [hunk 1]
```diff
@@ -263,9 +284,13 @@ static struct manifest *manifest_from_file(int version, char *component, bool he
 		} else if (c[0] == 'M') {
 			file->is_manifest = 1;
 		} else if (c[0] == 'I') {
-			/* ignore this file for future iterative manifest feature */
-			free(file);
-			continue;
+			/* only load iterative manifests when enabled*/
```
Comment (castulo) [inline src/manifest.c]: missing white space enabled*/

File: src/manifest.c [hunk 2]
```diff
@@ -704,10 +790,83 @@ static bool is_installed_and_verified(struct file *file)
...
+		file = list->data;
+		list = list->next;
+
+		install = false;
+
+		if (file->last_change > current_version) {
+			if (file->last_change == server->minversion) {
+				/* When a file without a content change is updated by a new minversion,
+				   don't include it in the update list. To keep consistent with the
+				   fallback update list, count as a changed file.*/
```
Comment (castulo) [inline src/manifest.c]: another missing white space here -> file.*/

File: src/manifest.c [hunk 3]
```diff
@@ -1300,6 +1497,61 @@ int enforce_compliant_manifest(struct file **a, struct file **b, int searchsize,
 	return ret; // If collisions were found, so manifest is purely additive
 }
 
+/* The function splits a list of manifests into two lists:
+ * - full_manifests: this list will contain all the full manifests that were in the manifests list
+ * - consolidated_manifests: this list will contain all the valid iterative manifests from the manifests
+ *   list, plus the full manifest of those bundles that didn't have a valid iterative manifest
+ *   A valid iterative manifest is one that can be used during an update, meaning that the iterative
+ *   manifest must match the "from version" provided, and that it belongs to an already installed bundle*/
```
Comment (castulo) [inline src/manifest.c]: one more missing white space

File: src/subscriptions.c [hunk 4]
```diff
@@ -176,3 +176,38 @@ void create_and_append_subscription(struct list **subs, const char *component)
...
+				if (manifest->last_change > current_version) {
+					*update_subs = list_prepend_data(*update_subs, sub);
+				} else {
+					*nonupdate_subs = list_prepend_data(*nonupdate_subs, sub);
+				}
+				break;
+			}
+		}
+
+		/* subscribed bundle not found in server MoM, so its entry was removed*/
```
Comment (castulo) [inline src/subscriptions.c]: one more

File: test/functional/bundleadd/add-with-iterative.bats [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	create_bundle -n test-bundle-1 -f /foo,/baz/test-file1 "$TEST_NAME"
+	create_bundle -n test-bundle-2 -f /bar,/baz/test-file2 "$TEST_NAME"
+	create_version "$TEST_NAME" 20 10
+	update_bundle -i "$TEST_NAME" test-bundle-1 --update /baz/test-file1
+	create_version "$TEST_NAME" 30 20
+	update_bundle -i "$TEST_NAME" test-bundle-2 --delete /baz/test-file2
+
+}
+
+@test "bundle-add add bundle works fine when there are iterative manifests in other versions" {
```
Comment (castulo) [inline test/functional/bundleadd/add-with-iterative.bats]: All these tests are going to need to have test IDs added to them

File: src/update.c [hunk 6]
```diff
@@ -476,7 +474,17 @@ static int main_update()
 
 	timelist_timer_start(global_times, "Create Update List");
 
-	updates = create_update_list(server_manifest);
+	/* When the minversion is 0 (set to 0 when missing from MoM header),
+	   delta manifests will not be used and the old update list implementation
+	   will be used. When using the new update list implementation if a
+	   minversion update occurred without a minversion header, all files without
+	   a content change would download fullfiles. */
+	if (server_manifest->minversion == 0) {
```
Comment (jwakre) [inline src/update.c]: Mixer has included the minversion header field in the MoM for a while now. Are there any arguments for keeping this special case or can I consolidate these?

Comment (jwakre) [reply src/update.c]: If we can remove the call to link_manifests, update performance could be improved and several delta manifest corner cases could be eliminated.

**Comment by jwakre** [comment]
Since manifest deltas are re-enabled, this feature may no longer be desired. Closing PR.

**Pull request closed without merging**